status/status.go

```diff
@@ -128,3 +130,39 @@ func FromError(err error) (s *Status, ok bool) {
 	}
 	return nil, false
 }
+
+// WithDetails returns a new status with the provided details messages appended to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.Proto() != nil.
+	p := s.Proto()
+	for _, detail := range details {
+		any, err := ptypes.MarshalAny(detail)
+		if err != nil {
+			return nil, err
+		}
+		p.Details = append(p.Details, any)
+	}
+	return &Status{s: p}, nil
+}
+
+// Details returns a slice of details messages attached to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) Details() ([]proto.Message, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.s != nil.
+	details := make([]proto.Message, 0, len(s.s.Details))
+	for _, any := range s.s.Details {
+		detail := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+			return nil, err
+		}
```

Comment:
I agree we should return all the ones that we can process without error. The question is what we do with those that result in error?

I see two points:
1. That is going to be a extremely uncommon thing in practice, so we should make design such that it is easy and clean to use in the >99% of cases where there is no error. That makes me lean against 1.

Between 2 and 3, I think 2 would be my preference, returning []interface{} is not very appealing, it doesn't really tell you what you got. It is true though that in both 2 and 3 we will be returning an interface that needs to be type switched, so it is not that bad.

Personally, when returning an error, I ask myself how that error would be used by the user. I generally feel that there are few cases where the users care about multiple errors or handle them. That is even more true in this context where they would be already handling a status error, and handling the errors occurring during getting the status error's details doesn't sound very appealing, my gut says >99.99% of use cases they would just ignore these. So it is additional complexity in the face of use cases that are going to be very uncommon. But this is a general library so they will be users that would be in that <0.01%. And since the common way to use the details will be switching then maybe 3 is indeed not that bad. What we lose compared to 2 is that we cannot assume that the returned result is a proto.Message, that complicates the cases where a user would want to use these details without type switching, but again that is probably not going to be common case and they can still do what they want with a bit more work. So between these 3 and despite my personal dislike to return multiple errors from a function I think 3 is probably the better choice between these.